core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -161,21 +175,120 @@ public List<Snapshot> apply() {
+      idsToRetain.add(snapshotId);
+    }
+
+    for (long idToRemove : idsToRemove) {
+      List<String> refsForId = retainedIdToRefs.get(idToRemove);
+      Preconditions.checkArgument(refsForId == null,
+          "Cannot expire %s. Still referenced by refs: %s", idToRemove, refsForId);
+    }
+
+    Set<Long> branchSnapshotsToRetain = computeAllBranchSnapshotsToRetain(retainedRefs.values());
+    idsToRetain.addAll(branchSnapshotsToRetain);
+    TableMetadata.Builder updatedMetaBuilder = TableMetadata.buildFrom(base);
+
+    base.snapshots().stream()
+        .map(Snapshot::snapshotId)
+        .filter(snapshot -> !idsToRetain.contains(snapshot))
+        .forEach(idsToRemove::add);
+    updatedMetaBuilder.removeSnapshots(idsToRemove);
+
+    base.refs().keySet().stream()
+        .filter(ref -> !retainedRefs.containsKey(ref))
+        .forEach(updatedMetaBuilder::removeRef);
+
+    return updatedMetaBuilder.build();
+  }
+
+  private Map<String, SnapshotRef> computeRetainedRefs(Map<String, SnapshotRef> refs) {
+    Map<String, SnapshotRef> retainedRefs = Maps.newHashMap();
+    for (Map.Entry<String, SnapshotRef> refEntry : refs.entrySet()) {
+      String name = refEntry.getKey();
+      SnapshotRef ref = refEntry.getValue();
+      if (name.equals(SnapshotRef.MAIN_BRANCH)) {
+        retainedRefs.put(name, ref);
+        continue;
+      }
+
+      Snapshot snapshot = base.snapshot(ref.snapshotId());
+      long maxRefAgeMs = ref.maxRefAgeMs() != null ? ref.maxRefAgeMs() : defaultMaxRefAgeMs;
+      if (snapshot != null) {
+        long refAgeMs = now - snapshot.timestampMillis();
+        if (refAgeMs <= maxRefAgeMs) {
+          retainedRefs.put(name, ref);
+        }
+      } else {
+        LOG.warn("Removing invalid ref {}: snapshot {} does not exist", name, ref.snapshotId());
+      }
+    }
+
+    return retainedRefs;
+  }
+
+  private Set<Long> computeAllBranchSnapshotsToRetain(Collection<SnapshotRef> refs) {
+    Set<Long> branchSnapshotsToRetain = Sets.newHashSet();
+    for (SnapshotRef ref : refs) {
+      if (ref.isBranch()) {
+        long expireSnapshotsOlderThan = ref.maxSnapshotAgeMs() != null ? now - ref.maxSnapshotAgeMs() :
+            defaultExpireOlderThan;
+        int minSnapshotsToKeep = ref.minSnapshotsToKeep() != null ? ref.minSnapshotsToKeep() :
+            defaultMinNumSnapshots;
+        branchSnapshotsToRetain.addAll(
+            computeBranchSnapshotsToRetain(ref.snapshotId(), expireSnapshotsOlderThan, minSnapshotsToKeep));
+      }
+    }
+
+    return branchSnapshotsToRetain;
+  }
+
+  private Set<Long> computeDanglingSnapshotsToRetain(Collection<SnapshotRef> refs) {
+    Set<Long> snapshotsPartOfBranches = Sets.newHashSet();
+    for (SnapshotRef ref : refs) {
+      if (ref.isBranch()) {
+        Iterable<Snapshot> branchSnapshots = SnapshotUtil.ancestorsOf(ref.snapshotId(), base::snapshot);
+        branchSnapshots.forEach(snapshot -> snapshotsPartOfBranches.add(snapshot.snapshotId()));
```

Comment:
We can be more efficient (time wise, at the tradeoff of memory) by maintaining the mapping between a branch and all the ancestors for the branch so that it can be reused across both the dangling snapshots and the branch snapshots to retain computation. Otherwise we have to get the ancestors multiple times.